source/branching.rst

```diff
@@ -0,0 +1,210 @@
+    in the merge commit message which would otherwise be lost when the branch is deleted and
+    replicates the merge behavior if the PR is merged in the Github UI. However, some projects might
+    prefer to keep a cleaner history by excluding the merge commits. That is ok as well as the developers
+    are in agreement and understand the trade-offs.
+
+Branch Lifespans
+-----------------
+
+Recall that ``origin/develop`` and ``origin/master`` are the only long term branches. Every other
+branch should be as short-lived as possible while still accomplishing its purpose. Dead branches
+(exploratory, or failed attempts to address an issue), or feature branches that drag on a long time,
+cloud the picture for other developers. Since this feature branch is no longer needed, it can be deleted
+via the "delete this branch" button on GitHub, or via the command line::
+
+    $ git branch -d 42-my-new-feature
+
+
+Hotfix Branches
+---------------
+
+A hotfix branch is an emergency branch mandated by a critical bug or security flaw in the production
+release. Hotfix branches always derive from ``origin/master``, and eventually merge back into
+``develop`` and ``master``::
+
+    $ git checkout -b hotfix-2.0.1 master
+    # do the work neccessary to bump version numbers
+    $ git commit -a -m "Bumped version number to 2.0.1"
+
+When the patch or bug fix has been applied to the code, you are ready to put it into ``master``::
+
+    $ git commit -m "Fixed severe production problem"
+    $ git checkout master
+    $ git merge --no-ff hotfix-2.0.1
+    $ git tag -a 2.0.1
+    $ git checkout develop
+    $ git merge --no-ff hotfix-2.0.1
+    $ git branch -d hotfix-2.0.1
+
+.. note::
+
+    As previously noted the ``--no-ff`` flag is not required but is recommended. See the previous
+    note for more information on the advantages of using this flag.
+
+
+Production Releases
+-------------------
+
+When a new production release is ready to be deployed, the latest work from ``develop`` is advanced
+to the ``master`` branch. It is recommended that releases are tagged. Tags might denote the version
+number for the release or the date of the release. Prior to tagging you ensure that all changes on
+master are included in develop::
+
+    $ git checkout master
+    $ git pull origin master
+    $ git checkout develop
+    $ git pull origin develop
+    $ git merge master
+
+This merge should be a no-op. If there are changes which were not previously merged to develop you need
+to ensure that develop is still running as expected before preceeding. While on the ``develop`` branch
+you should 
+
+- Update change log with release notes
+- Update internal version numbers if tracked in the code
+- Run test suite
+
+Changes between each release should be noted in a ``CHANGELOG`` or ``CHANGES`` file stored at the root of the repository.
+The format of this file is flexible. You may look at https://github.com/caktus/margarita/blob/develop/CHANGES.rst
```

Comment:
or if the project has Sphinx docs, we often will document releases in a file like `docs/changes.rst` or `docs/release_notes.rst`.